Add docs for rolling out changes to RabbitMQ (#322)
stackhpc/stackhpc-kayobe-config: doc/source/operations/rabbitmq.rst

========
RabbitMQ
========

High Availability
=================

In order to improve the stability of RabbitMQ, some changes need to be rolled,
out. These changes are:

* Update RabbitMQ to version 3.9.22, if you are running old images on Wallaby
  or Xena, by synchronising and then pulling a new RabbitMQ container from
  Pulp.
* Enable the high availability setting via Kolla-Ansible
  ``om_enable_rabbitmq_high_availability``.

By default in Kolla-Ansible, two key options for the high availability of
RabbitMQ are disabled. These are durable queues, where messages are persisted
to disk; and classic queue mirroring, where messages are replicated across

> multiple exchanges. Without these, a deployment will be a poor state for
> managing any updates for RabbitMQ, or any outages that cause it to be stopped.

Reviewer: Might need to be a bit more diplomatic here.

```suggestion
multiple exchanges. Without these, a deployment has a higher risk of experiencing
issues when updating RabbitMQ, or recovering from network outages.
```

Messages held in RabbitMQ nodes that are stopped will be lost, which causes
knock-on effects to the OpenStack services which either sent or were expecting
to receive them. The Kolla-Ansible flag
``om_enable_rabbitmq_high_availability`` can be used to enable both of these

> options. This will default to ``true`` from Xena onwards.

Reviewer: ```suggestion
options. The default will be overridden to ``true`` from Xena onwards in StackHPC Kayobe configuration.
```

While the `RabbitMQ docs <https://www.rabbitmq.com/queues.html#durability>`_ do
say "throughput and latency of a queue is not affected by whether a queue is
durable or not in most cases", it should be mentioned that there could be a
potential performance hit from replicating all messages to the disk within
large deployments. These changes would therefore be a tradeoff of performance
for stability.

**NOTE:** There is guaranteed to be downtime during this procedure, as it
requires restarting RabbitMQ and all the OpenStack services that use it. The
state of RabbitMQ will also be reset.

Instructions
------------

> 

Reviewer: ```suggestion
If you are planning to perform an upgrade, it is recommended to first roll out these changes.
```

> If you are upgrading from Wallaby to Xena, or if you're on Wallaby and want
> RabbitMQ to be more stable, you will need to enable the HA config option in

Reviewer: ```suggestion
If you are currently running Wallaby, you will need to enable the HA config option in
```

``etc/kayobe/kolla/globals.yml``.

.. code-block:: console

  om_enable_rabbitmq_high_availability: true

> Synchronise the Pulp containers.

Reviewer: ```suggestion
If you are running Wallaby or Xena, synchronise the Pulp containers.
```

.. code-block:: console

  kayobe playbook run etc/kayobe/ansible/pulp-container-sync.yml pulp-container-publish.yml -e stackhpc_pulp_images_kolla_filter=rabbitmq

Generate the new config files for the overcloud services.

.. code-block:: console

  kayobe overcloud service configuration generate

Pull the RabbitMQ container image.

.. code-block:: console

  kayobe overcloud container image pull -kt rabbitmq

Stop all the OpenStack services which use RabbitMQ.

.. code-block:: console

  kayobe overcloud host command run --command "docker ps -a | egrep '(barbican|blazar|ceilometer|cinder|cloudkitty|designate|heat|ironic|keystone|magnum|manila|masakari|neutron|nova|octavia)' | awk '{ print $NF }' | xargs docker stop"

Upgrade RabbitMQ.

.. code-block:: console

  kayobe overcloud service upgrade -kt rabbitmq --skip-prechecks

In order to convert the queues to be durable, you will need to reset the state
of RabbitMQ, and restart the services which use it. This can be done with the
RabbitMQ hammer playbook:

.. code-block:: console

>   kayobe playbook run stackhpc-kayobe-config/etc/kayobe/ansible/rabbitmq-reset.yml

Reviewer: Does this playbook restart all of the services we stopped?

Author: Oh, no it just restarts the services that have been seen to break when RabbitMQ goes down. IMO it's best to keep the playbook only targeting this subset so it can be used separately if rabbit breaks. So I'll add the step to start all the services back into the instructions again.

===============================

> Rolling out changes to RabbitMQ

Reviewer: Could just be `RabbitMQ`, with a subsection for `High Availability`

===============================

> Prior to upgrading, some changes need to be rolled out for RabbitMQ.

Reviewer: Could do with a bit more background here. e.g.

* what you get in kolla-ansible by default & potential problems with it
* what you get when the flag is set
* when we change the default (xena)

Then move onto how to roll it out, with conditions for different starting versions

Author: I've added some background here, would like to know if you think it should go more in-depth on the potential problems?

This should make RabbitMQ more stable during the upgrade(s). These changes
are:

* Enable the high availability setting via Kolla-Ansible

>   ``om_enable_rabbitmq_high_availability``.

Reviewer 2: Should we mention the potential for a performance hit here? Perhaps only some sites with slow disks / controllers will be affected. I suppose we won't be doing them all at once, so we can gather data as we go.

You would imagine that there would be significant overhead replicating everything to disk at large deployments.

Reviewer: RabbitMQ docs claim otherwise, but it's hard to know for sure without trying. It's worth calling out.

Author: I've added a bit to mention the potential performance hit, it does seem like a good idea to come back and update this once we've tried it out on production systems.

> * Update RabbitMQ to version 3.9.22

Reviewer: Need a bit more info on when to do this (wallaby & xena if running old images), and how to do so.

**NOTE:** There is guaranteed to be downtime during this procedure, as it
requires restarting RabbitMQ and all the OpenStack services that use it. The

> state of RabbitMQ may also need to be reset.

Reviewer 2: Is it a necessary step to flatten Rabbit so that all of the long-lived queues are converted to durable? I'm wondering if we should always do it.

Instructions
------------

> If you are upgrading from Wallaby to Xena, you will need to enable the HA

Reviewer: Or even if you're on Wallaby and want a more stable Rabbit

config option in

> ``/home/ubuntu/kayobe/config/src/kolla-ansible/ansible/group_vars/all.yml``.

Reviewer: This is the Kolla Ansible source code which should not be modified (and is an absolute path which won't always be valid). Instead, add the variable in `etc/kayobe/kolla/globals.yml`.

This will default to ``true`` from Xena onwards.

.. code-block:: console

  om_enable_rabbitmq_high_availability: true

Synchronise the Pulp containers.

.. code-block:: console

  kayobe playbook run etc/kayobe/ansible/pulp-container-sync.yml pulp-container-publish.yml -e stackhpc_pulp_images_kolla_filter=rabbitmq

Generate the new config files for the overcloud services.

.. code-block:: console

>   kayobe overcloud servuce configuration generate

Reviewer 2: nit: s/servuce/service

Pull the RabbitMQ container image.

.. code-block:: console

  kayobe overcloud container image pull -kt rabbitmq

Stop all the OpenStack services which use RabbitMQ.

.. code-block:: console

>   kayobe overcloud host command run --command 'docker ps -a | egrep '(barbican|blazar|ceilometer|cinder|cloudkitty|designate|heat|ironic|keystone|magnum|manila|masakari|neutron|nova|octavia)' | awk '{ print $NF }' | xargs docker stop'

Reviewer: We've got some nested quote issues here

Author: I've changed the outer quotes to double quotes. This step needs to be looked into further, as the formatting of the output is not great when running the command through kayobe.

Upgrade RabbitMQ.

.. code-block:: console

  kayobe overcloud service upgrade -kt rabbitmq --skip-prechecks

Check to see if RabbitMQ is functioning as expected.

.. code-block:: console

>   kayobe overcloud service upgrade -kt rabbitmq --skip-prechecks

Reviewer: Already did this :)

  kayobe overcloud host command run --command 'docker exec rabbitmq rabbitmqctl cluster_status'

>   kayobe overcloud host command run --command 'docker exec rabbitmq rabbitmqctl list_queues name durable'

Reviewer: Better add `--show-output` to these

The cluster status should list all controllers. The queues listed should be
durable if their names do not start with the following:

* amq.
* .\*\_fanout\_
* reply\_

> At this stage, you may find that the above is not correct. If this is the case,

Reviewer 2: As above, I'm wondering if we should /always/ do this. IIRC you need to re-declare queues to convert them to durable?

Reviewer: +1

Author: I've just tested this, and indeed we do need to reset rabbit in order to convert queues to durable. Just restarting the services is not enough

running the rabbitmq hammer playbook will reset the state of the RabbitMQ
nodes, and then bring the services which use RabbitMQ back up.

.. code-block:: console

  kayobe playbook run stackhpc-kayobe-config/etc/kayobe/ansible/rabbitmq-reset.yml

If you do not need to use the playbook here, then you will need to bring the
services back up yourself:

.. code-block:: console

  kayobe overcloud host command run --command 'docker ps -a | egrep '(barbican|blazar|ceilometer|cinder|cloudkitty|designate|heat|ironic|keystone|magnum|manila|masakari|neutron|nova|octavia)' | awk '{ print $NF }' | xargs docker start'

If there are issues with the services after this, you may still need to run the

> hammer playbook.

Reviewer 2: /me shudders!

Reviewer: :hammer: :rabbit:
